### ISSUE TITLE
libethash-cl is supposed to support OpenCL 1.1 but code uses 1.2 specific features

### ISSUE THREAD

**Issue by faenil** [ethereum/cpp-ethereum#3020]
According to https://github.com/ethereum/libethereum/blob/develop/libethash-cl/ethash_cl_miner.cpp#L354
the libethash-cl is supposed to support OpenCL 1.1, although the .cl code uses the "static" keyword multiple times.

The static keyword has been introduced in OpenCL 1.2, so it looks to me like it is not 1.1 compatible (anymore?) :)

I hit this blocker while trying to get ethminer working on Mesa+libclc + radeon OSS driver, which is supposedly an OpenCL 1.1 compatible setup :) 
It would be great if you could make the code 1.1 compatible so that users will be able to mine on OSS drivers :) (on a limited number of GPU architectures)

**Comment by bobsummerwill** [comment]
Hey @faenil,

If you have changes to suggest to meet this need, please do submit PRs.    I doubt there would be any objection from anybody.    I personally know close to nothing about OpenCL, and wouldn't know what changes would be needed :-)

You could also look at http://github.com/Genoil/cpp-ethereum, to see if Genoil's code is better on this score.

**Comment by faenil** [comment]
Hi @bobsummerwill :)

Unfortunately I'm in the same situation as you :/

I got the self-explanatory "OpenCL 1.1 does not support the static keyword" and went from there, with a bit of research...

**Comment by bobsummerwill** [comment]
You going to have a go at this, do you think, @faenil?    Feeling itchy? :-)

**Comment by faenil** [comment]
I tried, but got to a point where I get no errors but ethminer gets stuck at the point of starting the computation (i.e. after the first round of 0H/s). No idea where to go from there and I'm not willing to invest another weekend on it :)
It could be a problem anywhere in: Mesa, radeon driver, libclc, llvm, ethminer, libethereum...

**Comment by bobsummerwill** [comment]
Hey @Genoil,
Do you know of anybody working towards "Mesa+libclc + radeon OSS driver" ?   Best wishes!

**Comment by bobsummerwill** [comment]
Just going to cut-and-paste some further chat between @faenil and myself on IRC:

I spend the weekend trying to get it to run on latest mesa/libclc/llvm...but no luck, OpenCL1.1 is not enough
oh here's Bob :D

there could be multiple bugs at any point in that stack, as the OSS world is still OpenCL1.1 WIP, so...not really worth spending more time on
given that I have no experience with any of those pieces
s/spend/spent

alterego - So it's not mining properly still?

yes, with closed blobs, which I finally managed to get to work without X (as AMD dropped support for X 1.18 so Ubuntu dropped fglrx support)
but I had already tried 2-3 times before and failed, so I spent last weekend trying to get the OSS drivers to work with ethminer

tldr: 1) patch kernel module that does not otherwise build on 4.x (thank God someone had already done most of the job...I think it was https://bluehatrecord.wordpress.com/2016/03/25/installing-the-proprietary-amd-crimson-driver-on-fedora-23-with-linux-kernel-4-4-6-300/ even though last time I tried that post wasn't out yet, I used a previous one from the same guy in March), 2) uninstall any OpenCL-related Mesa package/file, as otherwise the system would always try to run Mesa, 3) build the AMD packages using --NoXServer (the web is full of people saying you can't do OpenCL without X on AMD, and then today I found this flag...), 4) copy amdpcsdb.default as suggested by http://www2.ati.com/relnotes/amd-catalyst-graphics-driver-installer-notes-for-linux-operating-systems.pdf

bob - Oh, I see.    So you have to disable OpenCL right now?

I have to uninstall any OpenCL-Mesa relate pkg, so that it uses AMD -OpenCL

bob - Right. And that one closed source one does support OpenCL 1.2, so you don't see the issue there?

M-bobsummerwill1: correct, it supports OpenCL2.0
M-bobsummerwill1: I already had everything working on Ubuntu 15.10 since March, but I wanted 16.04, because of its features. And 16.04 ships X 1.18, which is incompatible with fglrx AMD driver. So I managed to run fglrx with no X interactions, so it doesn't crash, and I can use fglrx only for OpenCL, on 16.04
if you install fglrx as AMD ships it, your system won't boot anymore. You need the steps I posted above to get something booting (using a secondary non-AMD card, in my case the server mobo iGPU)
I'm so happy :)
* faenil (IRC) goes to bed with a smile
let's hope they don't fry the GPUs, since without X I can't check the temperatures :P

**Comment by lvella** [comment]
Hey. I just stumbled (again) on this issue. Last time I tried to mine ether on Mesa + Free Drivers, they had a memory limitation that allowed at most 1GB to allocated on the GPU, and I couldn't make it work. Now it seems that this particular limitation is gone, and if I remember correctly, to fix the `static` keyword issue was a simple matter of adding the following line to the top of the OpenCL source file:

`
#pragma OPENCL EXTENSION cl_clang_storage_class_specifiers : enable
`

That enables the OpenCL 1.1 extension that allows for the use of the `static` keyword. [Intel from here.](https://llvm.org/svn/llvm-project/cfe/trunk/test/Parser/opencl-storage-class.cl) I am just now recompiling ethereum-cpp with the aforementioned change to the file "cpp-ethereum/libethash-cl/ethash_cl_miner_kernel.cl". I'll report back here if it works.

**Comment by faenil** [comment]
Looks like you are :) with a lot of calm, but you are :D

**Comment by pelelademadera** [comment]
for me, not working...

**Comment by EUA** [comment]
Not working here also on HD6870
```
Genoil's ethminer 0.9.41-genoil-1.1.7
=====================================================================
Forked from github.com/ethereum/cpp-ethereum
CUDA kernel ported from Tim Hughes' OpenCL kernel
With contributions from nicehash, nerdralph, RoBiK and sp_ 

Please consider a donation to:
ETH: 0xeb9310b185455f863f526dab3d245809f6854b4d

[OPENCL]:Found suitable OpenCL device [AMD BARTS (DRM 2.49.0 / 4.10.11-1-ARCH, LLVM 4.0.0)] with 1073741824 bytes of GPU memory
  m  00:17:04|ethminer  Getting work package...
JSON-RPC problem. Probably couldn't connect. Retrying in 1... 
  m  00:17:06|ethminer  Getting work package...
  m  00:17:06|ethminer  Got work package: #036a1b22
  ℹ  00:17:06|openclminer0  set work; seed: #00000000, target:  #000000004008
  ℹ  00:17:06|openclminer0  Initialising miner...
  m  00:17:06|ethminer  Mining on PoWhash #036a1b22 : 0.00MH/s [A0+0:R0+0:F0]
[OPENCL]:Using platform: Clover
[OPENCL]:Using device: AMD BARTS (DRM 2.49.0 / 4.10.11-1-ARCH, LLVM 4.0.0)(OpenCL 1.1 Mesa 17.0.4)
  m  00:17:07|ethminer  Mining on PoWhash #036a1b22 : 0.00MH/s [A0+0:R0+0:F0]
LLVM ERROR: Cannot select: 0x7f03680764e8: i32,ch = AtomicLoadAdd<Volatile LDST4[%g_output(addrspace=1)]> 0x7f036882a5b0, 0x7f0368099460, Constant:i32<1>
  0x7f0368099460: i32,ch = CopyFromReg 0x7f036882a5b0, Register:i32 %vreg186
    0x7f0368099530: i32 = Register %vreg186
  0x7f0368076b38: i32 = Constant<1>
In function: ethash_search

```

**Comment by darkicekeel** [comment]
my actual OS config
  OPERATING SYSTEM:   Gentoo 2.3
    Kernel:           4.9.16-gentoo (x86_64)
    Desktop:          KDE Frameworks 5
    Display Server:   X Server 1.19.2
    Compiler:         GCC 4.9.4 + Clang 5.0.0 + LLVM 5.0.0svn

cpp-ethereum-master/ethminer output:

[OPENCL]:Found suitable OpenCL device [AMD TAHITI (DRM 3.8.0 / 4.9.16-gentoo, LLVM 4.0.0)] with 4164202496 bytes of GPU memory
  ℹ  13:10:22|ethminer  Connecting to stratum server coinotron.com:3344
  ℹ  13:10:22|stratum  Connected to stratum server coinotron.com : 3344
  ℹ  13:10:22|stratum  Starting farm
  ℹ  13:10:22|stratum  Subscribed to stratum server
  ℹ  13:10:22|stratum  Received new job #403
  ℹ  13:10:22|openclminer0  set work; seed: #647f6619, target:  #000000010000
  ℹ  13:10:22|openclminer0  Initialising miner...
  ℹ  13:10:22|stratum  Authorized worker darkice.eth002
  m  13:10:24|ethminer  Mining on PoWhash #fb8e1a48 : 0.00MH/s [A0+0:R0+0:F0]
[OPENCL]:Using platform: Clover
[OPENCL]:Using device: AMD TAHITI (DRM 3.8.0 / 4.9.16-gentoo, LLVM 4.0.0)(OpenCL 1.1 Mesa 17.1.1)
[OPENCL]:input.cl:97:1: error: OpenCL version 1.1 does not support the 'static' storage class specifier
input.cl:114:1: error: OpenCL version 1.1 does not support the 'static' storage class specifier
input.cl:123:1: error: OpenCL version 1.1 does not support the 'static' storage class specifier
input.cl:209:1: error: OpenCL version 1.1 does not support the 'static' storage class specifier
input.cl:240:1: error: OpenCL version 1.1 does not support the 'static' storage class specifier
input.cl:245:1: error: OpenCL version 1.1 does not support the 'static' storage class specifier
input.cl:250:1: error: OpenCL version 1.1 does not support the 'static' storage class specifier
input.cl:380:1: error: OpenCL version 1.1 does not support the 'static' storage class specifier
                                                                                                                                                                                                                                                                               
[OPENCL]:clEnqueueWriteBuffer(-38)                                                                                                                                                                                                                                             
  m  13:10:26|ethminer  Mining on PoWhash #fb8e1a48 : 0.00MH/s [A0+0:R0+0:F0]
[...]

**Comment by chfast** [comment]
I think you should report it in https://github.com/ethereum-mining/ethminer. Here there is not ethminer any more.

**Comment by JHGitty** [comment]
See https://github.com/ethereum-mining/ethminer/issues/59
